go/vt/vtgate/semantics/semantic_state.go

```diff
@@ -983,3 +991,39 @@ func (st *SemTable) NewTableId() TableSet {
 	st.Tables = append(st.Tables, nil)
 	return tableID
 }
+
+func (st *SemTable) CanTakeSelectUnshardedShortcut() (*vindexes.Keyspace, bool) {
+	return canTakeSelectUnshardedShortcut(st.Tables)
+}
+
+func (st *SemTable) CanTakeUnshardedShortcut() (*vindexes.Keyspace, bool) {
+	return canTakeUnshardedShortcut(st.Tables)
+}
+
+func canTakeUnshardedShortcut(tableInfos []TableInfo) (*vindexes.Keyspace, bool) {
+	uks, _ := singleUnshardedKeyspace(tableInfos)
+	return uks, uks != nil
+}
+
+func canTakeSelectUnshardedShortcut(tableInfos []TableInfo) (*vindexes.Keyspace, bool) {
+	if mi := mirrorInfo(tableInfos); mi.Percent > 0 {
+		return nil, false
+	}
+	return canTakeUnshardedShortcut(tableInfos)
+}
+
+func (st *SemTable) GetMirrorInfo() MirrorInfo {
+	return mirrorInfo(st.Tables)
+}
+
+func mirrorInfo(tableInfos []TableInfo) MirrorInfo {
+	mi := MirrorInfo{}
+	for _, t := range tableInfos {
+		if mr := t.GetMirrorRule(); mr != nil {
+			if mi.Percent == 0 || mr.Percent < mi.Percent {
+				mi.Percent = mr.Percent
+			}
+		}
+	}
+	return mi
+}
```

Comment:
nit: can we not return from inside the if condition when it becomes true?

Reply:
even the mirror rule can have 0, should we not check the percent on the table info to be non-zero?
why do we want to continue to find the maximum percent?

Reply:
we are looking for the minimum, non-zero percent because the engine primitive operates based on a single percentage, but individual rules can have different percentages. so if multiple mirror rules are involved in a query, i need to pick one. 

i think the minimum makes sense because if i have one mirror rule that is 100%, and another that is 5%, i would rather the 5% rule get used in case i am trying to avoid overwhelming queries to that table.

Reply:
Let add this as a comment, if not already done.